### PR TITLE
ENG-4664 implement fetching current tenant and dynamic resource urls

### PR DESCRIPTION
**Pull request by ichalagashvili**

### CODE REVIEW

File: src/hooks/useDynamicResourceUrl.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
+import { useSelector } from 'react-redux';
+import { getDomain } from 'helpers/resourcePath';
+
+import { selectCurrentTenant } from 'state/multi-tenancy/selectors';
+
+export const generateUrlFromTenantAndResource = ({ tenant, resource }) => {
+  console.log('generateUrlFromTenantAndResource', tenant, resource);
```
Comment (Dv92) [inline src/hooks/useDynamicResourceUrl.js]: can you delete this `console.log`?

File: src/hooks/useDynamicResourceUrl.js [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+  const resourceUrl = tenant.resourceRootUrl || '';
+  const resourcePath = tenant.resourceRootPath || '';
+  // used for mfe local development testing
+  if (process.env.USE_LOCAL_MFE) {
+    console.log('returning local resource', resource);
```
Comment (Dv92) [inline src/hooks/useDynamicResourceUrl.js]: same here

File: src/hooks/useDynamicResourceUrl.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    return resource;
+  }
+  if (process.env.USE_REMOTE_MFE) {
+    console.log('returning remote resource', `${getDomain()}/cmsresources/${resource}`);
```
Comment (Dv92) [inline src/hooks/useDynamicResourceUrl.js]: same here

File: src/hooks/useDynamicResourceUrl.js [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+    return `${getDomain()}/cmsresources/${resource}`;
+  }
+  if (resourceUrl) {
+    console.log('returning resourceUrl', `${resourceUrl}/${resource}`);
```
Comment (Dv92) [inline src/hooks/useDynamicResourceUrl.js]: same here

File: src/hooks/useDynamicResourceUrl.js [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+    return `${resourceUrl}/${resource}`;
+  }
+  if (resourcePath) {
+    // please not that getDomain is not used here since the subpath
+    // is already included in the resourcePath (e.g /entando-de-app)
+    console.log('returning resourcePath', `${resourcePath}${resource}`);
```
Comment (Dv92) [inline src/hooks/useDynamicResourceUrl.js]: same here

File: src/hooks/useDynamicResourceUrl.js [hunk 6]
```diff
@@ -0,0 +1,41 @@
...
+    return `${resourcePath}${resource}`;
+  }
+  // if you are using multipass on local instance then replace <resources> with <cmsresources>
+  const localResourceUrl = `${getDomain()}/resources/${resource}`;
+  console.log('logging localResourceUrl', localResourceUrl);
+  console.log('final output is: ', process.env.NODE_ENV === 'production' ? `${getDomain()}/cmsresources/${resource}` : localResourceUrl);
```
Comment (Dv92) [inline src/hooks/useDynamicResourceUrl.js]: same here

File: src/hooks/useDynamicResourceUrl.js [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+  return process.env.NODE_ENV === 'production' ? `${getDomain()}/cmsresources/${resource}` : localResourceUrl;
+};
+
+
+export function useDynamicResourceUrl(resource) {
+  const tenant = useSelector(selectCurrentTenant) || {};
+  console.log('useDynamicResourceUrl', tenant, resource);
```
Comment (Dv92) [inline src/hooks/useDynamicResourceUrl.js]: same here

File: src/hooks/useDynamicResourceUrl.js [hunk 8]
```diff
@@ -0,0 +1,41 @@
+import { useSelector } from 'react-redux';
```
Comment (Dv92) [inline src/hooks/useDynamicResourceUrl.js:1]: I don't see the test for `useDynamicResourceUrl`and `generateUrlFromTenantAndResource`

File: src/hooks/useDynamicResourceUrl.js [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+import { getDomain } from 'helpers/resourcePath';
+
+import { selectCurrentTenant } from 'state/multi-tenancy/selectors';
+
+export const generateUrlFromTenantAndResource = ({ tenant, resource }) => {
```
Comment (Dv92) [inline src/hooks/useDynamicResourceUrl.js]: The name of this function seems too long, what do you think of `generateStaticAssetsUrl`?

File: src/hooks/useMfe.js [hunk 10]
```diff
@@ -1,26 +1,19 @@
...
 // try to find the element by id
 const isJSLoaded = id => document.getElementById(`script-${id}`);
 const isCSSLoaded = id => document.getElementById(`style-${id}`);
 
 // generates the <script> tag and track the loading status
-const createScript = (id, asset, remove, setError) => {
+const createScript = (id, asset, remove, setError, tenant) => {
   const script = document.createElement('script');
-  script.src = getMfeResourcePath(asset);
+  console.log('script source is: ', tenant, asset, generateUrlFromTenantAndResource({ tenant, resource: asset }));
```
Comment (Dv92) [inline src/hooks/useMfe.js]: can you delete this `console.log`?

File: src/hooks/useMfe.js [hunk 11]
```diff
@@ -50,13 +44,14 @@ const createStyle = (id, asset, remove, setError) => {
 };
 
 // inject asset to DOM accordingly to type
-const injectAssetToDom = ({ id, assets = [] }, remove, setError) => {
+const injectAssetToDom = ({ id, assets = [] }, remove, setError, tenant) => {
+  console.log('injectAssetToDom', id, assets, tenant);
```
Comment (Dv92) [inline src/hooks/useMfe.js]: same here

File: src/hooks/useMfe.js [hunk 12]
```diff
@@ -33,12 +26,13 @@ const createScript = (id, asset, remove, setError) => {
...
 
 // generates the <style> tag and track the loading status
-const createStyle = (id, asset, remove, setError) => {
+const createStyle = (id, asset, remove, setError, tenant) => {
   const styles = document.createElement('link');
   styles.rel = 'stylesheet';
   styles.type = 'text/css';
   styles.media = 'screen';
-  styles.href = getMfeResourcePath(asset);
+  console.log('style source is: ', tenant, asset, generateUrlFromTenantAndResource({ tenant, resource: asset }));
```
Comment (Dv92) [inline src/hooks/useMfe.js]: same here

File: src/hooks/useMfe.js [hunk 13]
```diff
@@ -91,12 +89,15 @@ const useMfe = ({ mfeId, initialMfe }) => {
     setAssetLoading(prev => (prev || []).filter(a => a !== id));
   }, []);
 
+  console.log('useMfe', memoMfe, tenant, assetLoading, hasError);
```
Comment (Dv92) [inline src/hooks/useMfe.js]: same here

File: src/hooks/useMfe.js [hunk 14]
```diff
@@ -79,8 +74,11 @@ const deleteAssetFromDom = ({ id, assets = [] }) => {
   });
 };
 
+const EMPTY_OBJECT = {};
+
 const useMfe = ({ mfeId, initialMfe }) => {
   const mfe = useSelector(state => getMfeById(state, mfeId));
+  const tenant = useSelector(selectCurrentTenant) || EMPTY_OBJECT;
```
Comment (Dv92) [inline src/hooks/useMfe.js:78]: why do you need a`EMPTY_OBJECT` constant?

File: src/state/multi-tenancy/actions.js [hunk 15]
```diff
@@ -0,0 +1,24 @@
...
+    const json = await response.json();
+    if (response.ok && json.payload) {
+      dispatch(setCurrentTenant(json.payload));
+    } else {
+      // json.errors.forEach(err => dispatch(addToast(err.message, TOAST_ERROR)));
+      // do nothing...
+    }
+  } catch (e) {
+    // do nothing...
+  }
```
Comment (Dv92) [inline src/state/multi-tenancy/actions.js]: is missing error handling for this action

File: src/ui/app/MfeContainer.js [hunk 16]
```diff
@@ -6,15 +6,19 @@ import PropTypes from 'prop-types';
...
+import { useDynamicResourceUrl } from 'hooks/useDynamicResourceUrl';
 
 const MfeContainer = ({ id, history }) => {
   const { assetLoading, mfe } = useMfe({ mfeId: id });
   const locale = useSelector(getLocale);
   const permissions = useSelector(getLoggedUserPermissions);
   const systemReport = useSelector(getSystemReport);
 
+  const mfeResourceBasePath = useDynamicResourceUrl(mfe.assetsBasePath);
+  console.log('MfeContainer->mfeResourceBasePath', mfeResourceBasePath);
```
Comment (Dv92) [inline src/ui/app/MfeContainer.js]: can you delete this `console.log`?

File: src/ui/app/MfeContainer.js [hunk 17]
```diff
@@ -32,11 +36,13 @@ const MfeContainer = ({ id, history }) => {
     }
 
     entandoWindow.epc = entandoWindow.epc || {};
+    console.log('inside useEffect->mfeResourceBasePath', mfe.widgetName, mfeResourceBasePath, entandoWindow.epc[mfe.widgetName] || { basePath: mfeResourceBasePath });
```
Comment (Dv92) [inline src/ui/app/MfeContainer.js]: same here

File: src/ui/widgets/config/WidgetConfigMicrofrontend.js [hunk 18]
```diff
@@ -21,6 +21,7 @@ const WidgetConfigMicrofrontend = ({
   onSubmit, widget, widgetConfig, onCancel, widgetCode, framePos, frameName, pageCode,
 }) => {
   const resources = get(widget, 'configUi.resources', []);
+  console.log('inside WidgetConfigMicrofrontend resources', resources);
```
Comment (Dv92) [inline src/ui/widgets/config/WidgetConfigMicrofrontend.js]: same here

File: src/ui/widgets/config/WidgetConfigMicrofrontend.js [hunk 19]
```diff
@@ -55,6 +56,12 @@ const WidgetConfigMicrofrontend = ({
   const microfrontendMarkup = renderMicrofrontend(customElement);
   const shouldRender = !assetLoading && !hasError && resources.length && customElement;
 
+  console.log('inside WidgetConfigMicrofrontend shouldRender', shouldRender, assetLoading, hasError, resources.length, customElement, {
+    resources,
+    mfe,
+    microfrontendMarkup,
+  });
```
Comment (Dv92) [inline src/ui/widgets/config/WidgetConfigMicrofrontend.js]: same here

File: src/ui/widgets/config/WidgetConfigPage.js [hunk 20]
```diff
@@ -105,6 +105,7 @@ class WidgetConfigPage extends Component {
           </WidgetConfigPanel>
         );
       }
+      console.log('isMicrofrontendWidgetForm(widget)', isMicrofrontendWidgetForm(widget), widget);
```
Comment (Dv92) [inline src/ui/widgets/config/WidgetConfigPage.js]: same here

**Review by Dv92 — CHANGES_REQUESTED** [review]
There are some logs to delete, missing some tests and small modification on naming

**Review by Dv92 — DISMISSED** [review]
LGTM

**Review by Dv92 — APPROVED** [review]

**Review by DMinnai — APPROVED** [review]

**Pull request merged**